src/compiler/checker.rs

```diff
@@ -0,0 +1,363 @@
+                };
+            }
+            _ => unimplemented!(),
+        };
+    }
+
+    fn check_source_file(&self, source_file: &SourceFile) {
+        self.check_source_file_worker(source_file)
+    }
+
+    fn check_source_file_worker(&self, node: &SourceFile) {
+        if true {
+            for_each(&node.statements, |statement, _index| {
+                self.check_source_element(statement);
+                Option::<()>::None
+            });
+        }
+    }
+
+    pub fn get_diagnostics(&self, source_file: &SourceFile) -> Vec<Rc<Diagnostic>> {
+        self.get_diagnostics_worker(source_file)
+    }
+
+    fn get_diagnostics_worker(&self, source_file: &SourceFile) -> Vec<Rc<Diagnostic>> {
+        self.check_source_file(source_file);
+
+        let semantic_diagnostics = self
+            .diagnostics()
+            .read()
+            .unwrap()
+            .get_diagnostics(&source_file.file_name);
+
+        semantic_diagnostics
+    }
+
+    fn check_arithmetic_operand_type(
+        &self,
+        operand: /*&Node*/ &Expression,
+        type_: Rc<Type>,
+        diagnostic: &DiagnosticMessage,
+    ) -> bool {
+        if !self.is_type_assignable_to(type_, self.number_or_big_int_type()) {
+            self.error_and_maybe_suggest_await(operand, diagnostic);
+            return false;
+        }
+        true
+    }
+
+    fn check_prefix_unary_expression(&self, node: &PrefixUnaryExpression) -> Rc<Type> {
+        let operand_expression = match &*node.operand {
+            Node::Expression(expression) => expression,
+            _ => panic!("Expected Expression"),
+        };
+        let operand_type = self.check_expression(operand_expression);
+        match node.operator {
+            SyntaxKind::PlusPlusToken => {
+                self.check_arithmetic_operand_type(operand_expression, operand_type.clone(), &Diagnostics::An_arithmetic_operand_must_be_of_type_any_number_bigint_or_an_enum_type);
+                return self.get_unary_result_type(&operand_type);
+            }
+            _ => {
+                unimplemented!();
+            }
+        }
+    }
+
+    fn get_unary_result_type(&self, operand_type: &Type) -> Rc<Type> {
+        self.number_type()
+    }
+
+    fn check_expression(&self, node: &Expression) -> Rc<Type> {
+        self.check_expression_worker(node)
+    }
+
+    fn check_expression_worker(&self, node: &Expression) -> Rc<Type> {
+        match node {
+            Expression::TokenExpression(token_expression) => match token_expression.kind() {
+                SyntaxKind::TrueKeyword => {
+                    return self.true_type().clone();
+                }
+                _ => unimplemented!(),
+            },
+            Expression::PrefixUnaryExpression(prefix_unary_expression) => {
+                return self.check_prefix_unary_expression(prefix_unary_expression);
+            }
+            // Expression::BinaryExpression(binary_expression) => {
```

Comment:
As mentioned, didn't implement binary-expression type-checking first because there's a "trampolining" optimization happening under the hood